src/parser/Parser.ts

```diff
@@ -98,65 +100,58 @@ export class Parser {
      */
     public ast: Body;
 
-    //TODO remove this once we have verified all of the tests.
-    public statements: Statement[];
-
-    /**
-     * The list of diagnostics found during the parse process
-     */
-    public diagnostics: Diagnostic[];
-
-    /**
-     * The depth of the calls to function declarations. Helps some checks know if they are at the root or not.
-     */
-    private namespaceAndFunctionDepth: number;
-
-    /**
-     * The options used to parse the file
-     */
-    public options: ParseOptions;
-
-    private globalTerminators = [] as TokenKind[][];
-
-    /**
-     * All class statements defined in this file
-     */
-    public classStatements = [] as ClassStatement[];
+    public get statements() {
+        return this.ast.statements;
+    }
 
     /**
-     * All namespace statements defined in this file
+     * References for significant statements/expressions in the parser.
+     * These are initially extracted during parse-time to improve performance, but will also be dynamically regenerated if need be.
+     *
+     * If a plugin modifies the AST, then the plugin should call Parser#clearReferences() to force this list to refresh
      */
-    public namespaceStatements = [] as NamespaceStatement[];
+    public get references() {
+        //build the references object if it's missing.
+        if (!this._references) {
+            this._references = this.findReferences();
+        }
+        return this._references;
+    }
 
-    /**
-     * All function statements defined in this file
-     */
-    public functionStatements = [] as FunctionStatement[];
+    private _references: References = {
+        assignmentStatements: [],
+        classStatements: [],
+        functionStatements: [],
+        functionExpressions: [],
+        importStatements: [],
+        libraryStatements: [],
+        namespaceStatements: [],
+        newExpressions: []
+    };
 
     /**
-     * All function expressions defined in this file
+     * Clears the references. This should be called anytime the AST has been manipulated.
      */
-    public functionExpressions = [] as FunctionExpression[];
+    clearReferences() {
+        this._references = undefined;
```

Comment:
Should reset to an object with empty arrays IMHO. Maybe even better should be an `invalidateReferences` so you know they need to be refreshed but you don't lose the data.

Reply:
I don't see any value in keeping the old references around since `Parser.references` is a getter that auto-regenerates `references` if they are cleared. If a plugin author has modified the AST, then none of these references can be considered valid anymore anyway.

Reply:
Ok but still `invalidateReferences` would be more consistent.

Reply:
done.